Applications/spire/source/charting/chart_view.cpp

```diff
@@ -0,0 +1,158 @@
+#include "spire/charting/chart_view.hpp"
+#include <locale>
+#include <boost/date_time/posix_time/posix_time.hpp>
+#include <QFontMetrics>
+#include <QPainter>
+#include <QPaintEvent>
+#include "Nexus/Definitions/Money.hpp"
+#include "spire/spire/dimensions.hpp"
+
+using namespace boost::posix_time;
+using namespace Nexus;
+using namespace Spire;
+
+namespace {
+  QVariant to_variant(ChartValue::Type type, ChartValue value) {
+    if(type == ChartValue::Type::DURATION) {
+      return QVariant::fromValue(static_cast<time_duration>(value));
+    } else if(type == ChartValue::Type::MONEY) {
+      return QVariant::fromValue(static_cast<Money>(value));
+    } else if(type == ChartValue::Type::QUANTITY) {
+      return QVariant::fromValue(static_cast<Quantity>(value));
+    } else if(type == ChartValue::Type::TIMESTAMP) {
+      return QVariant::fromValue(static_cast<ptime>(value));
+    }
+    return QVariant();
+  }
+}
+
+ChartView::ChartView(ChartValue::Type x_axis_type,
+    ChartValue::Type y_axis_type, QWidget* parent)
+    : QWidget(parent),
+      m_x_axis_type(x_axis_type),
+      m_y_axis_type(y_axis_type),
+      m_timestamp_format("%H:%M:%S") {
+  setFocusPolicy(Qt::NoFocus);
+  m_label_font = QFont("Roboto");
+  m_label_font.setPixelSize(scale_height(10));
+  m_item_delegate = new CustomVariantItemDelegate(this);
+  m_x_axis_type = ChartValue::Type::TIMESTAMP;
+  m_y_axis_type = ChartValue::Type::MONEY;
+  auto current_time = boost::posix_time::second_clock::local_time();
+  auto bottom_right = ChartPoint(
+    ChartValue(current_time),
+    ChartValue(Nexus::Money(10)));
+  auto top_left = ChartPoint(
+    ChartValue(current_time - boost::posix_time::hours(1)),
+    ChartValue(Nexus::Money(10.10)));
+  set_region(top_left, bottom_right);
+}
+
+void ChartView::set_region(ChartPoint top_left, ChartPoint bottom_right) {
+  m_top_left = top_left;
+  m_bottom_right = bottom_right;
+  update();
+}
+
+void ChartView::paintEvent(QPaintEvent* event) {
+  auto painter = QPainter(this);
+  painter.setFont(m_label_font);
+  painter.setPen(Qt::white);
+  auto font_metrics = QFontMetrics(m_label_font);
+  auto x_origin = width() -
+    (font_metrics.width("M") *
+      QString::number(
+        static_cast<double>(
+          static_cast<Quantity>(ChartValue(m_top_left.m_y))),
```

Comment:
This needs to be revised, for one ```m_top_left.m_y``` is already a ```ChartValue```, so no need to explicitly invoke the ```ChartValue``` constructor on it. Second, there's no guarantee that it's a ```Quantity```, so the cast to ```Quantity``` could have invalid behavior.